File: src/main/scala/com/yogpc/qp/data/LootTableSerializeHelper.scala
package com.yogpc.qp.data

import cats.Eval
import com.google.gson.JsonElement
import com.yogpc.qp.machines.base.IEnchantableTile
import net.minecraft.block.Block
import net.minecraft.data.loot.BlockLootTables
import net.minecraft.loot.functions.ILootFunction
import net.minecraft.loot.{ConstantRange, LootParameterSets, LootPool, LootTable, LootTableManager, StandaloneLootEntry}
import net.minecraft.util.ResourceLocation

case class LootTableSerializeHelper(block: Block,
                                    functions: List[ILootFunction.IBuilder] = Nil)
  extends BlockLootTables with QuarryPlusDataProvider.DataBuilder {
  override def location: ResourceLocation = block.getRegistryName

  override def build: JsonElement = {
    val value: StandaloneLootEntry.Builder[_] = SerializeUtils.builder(this.block, this.functions)

    val builder = LootTable.builder().addLootPool(BlockLootTables.withSurvivesExplosion(this.block, LootPool.builder().rolls(ConstantRange.of(1)).addEntry(value)))
    builder.setParameterSet(LootParameterSets.BLOCK)

    LootTableManager.toJson(builder.build())
  }

  def dropWithEnchantments: LootTableSerializeHelper =
    this.copy(functions = LootTableSerializeHelper.dropFunction.value :: this.functions)
}

object LootTableSerializeHelper {
  final val dropFunction: Eval[ILootFunction.IBuilder] = Eval.later(IEnchantableTile.DropFunction.builder())

  def withDrop(block: Block): LootTableSerializeHelper = LootTableSerializeHelper(block)

  def withEnchantedDrop(block: Block): LootTableSerializeHelper = LootTableSerializeHelper(block, List(dropFunction.value))
}

File: src/main/scala/com/yogpc/qp/data/LootTableSerializeHelper.java
package com.yogpc.qp.data;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

import com.google.gson.JsonElement;
import com.yogpc.qp.machines.EnchantedLootFunction;
import net.minecraft.data.loot.BlockLoot;
import net.minecraft.resources.ResourceLocation;
import net.minecraft.world.level.block.Block;
import net.minecraft.world.level.storage.loot.LootPool;
import net.minecraft.world.level.storage.loot.LootTable;
import net.minecraft.world.level.storage.loot.LootTables;
import net.minecraft.world.level.storage.loot.entries.LootItem;
import net.minecraft.world.level.storage.loot.entries.LootPoolSingletonContainer;
import net.minecraft.world.level.storage.loot.functions.LootItemFunction;
import net.minecraft.world.level.storage.loot.parameters.LootContextParamSets;
import net.minecraft.world.level.storage.loot.providers.number.ConstantValue;

class LootTableSerializeHelper extends BlockLoot implements QuarryPlusDataProvider.DataBuilder {
    private final Block block;
    private final List<LootItemFunction.Builder> functions;

    LootTableSerializeHelper(Block block, List<LootItemFunction.Builder> functions) {
        this.block = block;
        this.functions = functions;
    }

    @Override
    public ResourceLocation location() {
        return block.getRegistryName();
    }

    @Override
    public JsonElement build() {
        LootPoolSingletonContainer.Builder<?> value = LootItem.lootTableItem(block);
        functions.forEach(value::apply);
        var builder = LootTable.lootTable()
            .withPool(BlockLoot.applyExplosionCondition(block, LootPool.lootPool().setRolls(ConstantValue.exactly(1)).add(value)));
        builder.setParamSet(LootContextParamSets.BLOCK);
        return LootTables.serialize(builder.build());
    }

    LootTableSerializeHelper add(LootItemFunction.Builder builder) {
        var copy = new ArrayList<>(functions);
        copy.add(builder);
        return new LootTableSerializeHelper(block, copy);
    }

    @Override
    public String toString() {
        return "LootTableSerializeHelper{" +
            "block=" + block +
            ", functions=" + functions +
            '}';
    }

    static LootTableSerializeHelper withDrop(Block block) {
        return new LootTableSerializeHelper(block, Collections.emptyList());
    }

    static LootTableSerializeHelper withEnchantedDrop(Block block) {
        return new LootTableSerializeHelper(block, Collections.singletonList(EnchantedLootFunction.builder()));
    }
}
